test/unit/finance/bill_variable_for_plan_changed_test.rb

```diff
@@ -3,49 +3,56 @@
 require 'test_helper'
 
 class Finance::BillVariableForPlanChangedTest < ActiveSupport::TestCase
-  attr_reader :contract, :account, :app_plan
+  attr_reader :contract, :app_plan
 
   setup do
     @contract = FactoryBot.build_stubbed(:contract)
-    @account  = FactoryBot.build_stubbed(:simple_account)
+    account = FactoryBot.build_stubbed(:simple_account)
     @app_plan = FactoryBot.build_stubbed(:application_plan)
 
-    @org_tz = ENV["TZ"]
+    contract.stubs(:provider_account).returns(account)
+    account.stubs(:provider_can_use?).with(:instant_bill_plan_change).returns(true)
   end
 
-  teardown do
-    ENV["TZ"] = @org_tz
+  test "bill for variable on first day" do
+    Timecop.travel(Time.now.beginning_of_month) # rubocop:disable Rails/TimeZone we don't use timezones in billing
+
+    contract.expects(:save).never
+    contract.notify_observers(:bill_variable_for_plan_changed, app_plan)
   end
 
-  test "bill for variable" do
-    Timecop.travel(15.days.ago) if Time.now.mday == 1 # rubocop:disable Rails/TimeZone we don't use timezones in billing
+  test "bill for variable before first day" do
+    Timecop.travel(Time.now.beginning_of_month - 15.days) # rubocop:disable Rails/TimeZone we don't use timezones in billing
```

Comment:
`bill for variable after first day` and `bill for variable before first day` is one and the same thing.

Just current month is different. So we don't need two tests to check what happens in the middle of the month.

Also like to run the test in different days of month to have chance to catch any other special cases. I'll leave Hery to advise whether to use this or leave `Timecop.travel(15.days.ago) if Time.now.mday == 1`. Maybe we can iterate over every day of the month so we have full coverage.

Reply:
Waiting every day of the month in order to possibly find a bug doesn't sound right, considering we can force the time using `Timecop`. I think there are 2 strong scenarios here:  when it is the first day and when it isn't.

The 3rd edge case is interaction between UTC and local time which is already covered.

And we probably should work a bit more with `variable_cost_paid_until`.

> Maybe we can iterate over every day of the month so we have full coverage

Although a bit overkill, it would make sense.

> bill for variable after first day and bill for variable before first day is one and the same thing.

You're right I was thinking about `variable_cost_paid_until` when I wrote this. It is definitely redundant.

Reply:
There is another special case which is the day at a previous plan change (this is also covered). We can never be sure whether some time a new special case can be introduced. We have a lot of uncontrolled variance in tests. Like time zones, hour of day, etc. So it is not really much different.

I think a loop from 2nd to last day of month would be a good options that would be best of both worlds.

Reply:
To settle it we can just put a fix date we know is an issue ;)